feat: add user invitation doctype & related public methods

EltonLobo07 opened the pull request:
Adds a doctype and a couple of helper methods to store all of the common logic of a typical user invitation flow.

Until now, there was no way to invite and create a new user based on a sent invitation that can be accepted or rejected by the invitee. Due to this, custom Framework applications have to implement a user invitation flow. But most of the rules around this flow are generic enough to let Framework store all of the common logic associated with a typical user invitation flow. This will help ensure consistency and prevent code duplication for custom Framework applications that need this type of feature.

For more information: [internal doc](https://github.com/frappe/frappe/blob/44256c2796eb03abeaf1f9aefaadd81dac869638/frappe/core/doctype/user_invitation/internal_doc/index.md)

`no-docs`

EltonLobo07 committed:
feat: add user invitation doctype & related public methods

EltonLobo07 committed:
Merge branch 'develop' of https://github.com/frappe/frappe into feat-user-invitation

EltonLobo07 committed:
style(user-invitation): execute formatters & add semgrep comments

EltonLobo07 committed:
refactor(user-invitation): use `is` to compare `None` values

EltonLobo07 committed:
fix(user-invitation): skip fetching `after_accept` for default app

EltonLobo07 committed:
fix(user-invitation): translate email templates

EltonLobo07 committed:
fix(user-invitaton): return pending invites from invite by email api

ankush commented on frappe/core/api/user_invitation.py:
```diff
@@ -0,0 +1,110 @@
+def invite_by_email(
+	emails: str, role: str, redirect_to_path: str, app_name: str = FRAMEWORK_APP_NAME
+) -> dict[str, list[str]]:
+	frappe.only_for(["System Manager", "User Invitation Manager"])
```
`User Manager` is a better name.

ankush commented on frappe/core/api/user_invitation.py:
```diff
@@ -0,0 +1,110 @@
+	allowed_roles: list[str] = []
+	if not is_app_framework:
+		# `role` validation
+		invalid_app_msg = frappe._("Invalid app")
```
1. Prefer importing `_` separately. Improves readability. 
2. Write better error message. Use different message for different failures. This doesn't explain anything. 
3. Add title to error message. 

https://gameplan.frappe.cloud/g/space/94/discussion/2114/gyan-write-better-error-messages

ankush commented on frappe/core/api/user_invitation.py:
```diff
@@ -0,0 +1,110 @@
+	invalid_or_expired_key_msg = frappe._("Invalid or expired key")
+	if not result:
+		frappe.throw(invalid_or_expired_key_msg)
+	invitation = frappe.get_doc(USER_INVITATION_DOCTYPE, result[0])
```
It's okay to not use variable name for doctypes. Keep it consistent with rest of codebase where doctype name is repeated wherever it's used. Better readability IMO.

ankush commented on frappe/core/api/user_invitation.py:
```diff
@@ -0,0 +1,110 @@
+			allowed_roles = user_invitation_hook.get(allowed_invite_roles_key)
+			if not allowed_roles:
+				frappe.throw(invalid_app_msg)
+		except ImportError:
+			frappe.throw(invalid_app_msg)
```
Don't do this, first validate if app is even installed.

ankush commented on frappe/core/api/user_invitation.py:
```diff
@@ -0,0 +1,110 @@
+
+@frappe.whitelist()
+def invite_by_email(
+	emails: str, role: str, redirect_to_path: str, app_name: str = FRAMEWORK_APP_NAME
```
Avoid use of `" "` as sentinal value for default. You can use `None` as sentinal value here?

ankush commented on frappe/core/api/user_invitation.py:
```diff
@@ -0,0 +1,110 @@
+	)
+	to_invite = list(set(email_list) - set(existing_user_emails) - set(existing_invited_emails))
+	for email in to_invite:
+		frappe.get_doc(
```
Use `frapp.new_doc`

ankush commented on frappe/core/doctype/user_invitation/test_user_invitation.py:
```diff
@@ -0,0 +1,184 @@
+			email=emails[0],
+			role="System Manager",
+			redirect_to_path="/abc",
+			app_name=" ",  # " " represents Framework
```
Why not just call it `frappe`? That's the name of the framework "app".

ankush replied:
Hint: Often things you have to explain in comments are areas of improvement.

ankush reviewed (changes requested):
Code quality needs some work. Try to keep functions short and code layout "flatter".

(haven't reviewed functionality etc yet, will do after changes)

EltonLobo07 committed:
refactor(user-invitation): improve code quality

EltonLobo07 committed:
fix(user-invitation): translate all error messages

ankush commented:
Support adding more than one role.  <img width="746" height="377" alt="image" src="https://github.com/user-attachments/assets/2ff0dcd6-7023-456d-a4e8-7f237384cbb9" />

ankush commented on frappe/core/doctype/user_invitation/user_invitation.py:
```diff
@@ -0,0 +1,93 @@
+
+	def before_insert(self):
+		frappe.utils.validate_email_address(self.email, throw=True)
+		self.key = frappe.generate_hash(length=12)
```
This should be hashed for the sake of security, and the original key shouldn't be stored anywhere.

You can refer how framework handles reset password. https://github.com/frappe/frappe/pull/24857

ankush commented on frappe/core/doctype/user_invitation/user_invitation.py:
```diff
@@ -0,0 +1,93 @@
+				subject=f"Invitation to join {email_title} revoked",
+				template="user_invitation_revoked",
+				args={"title": email_title, "site_name": self.get_site_name()},
+				now=True,
```
Note to self: We should have something between instantly and via queue.

ankush commented on frappe/core/doctype/user_invitation/user_invitation.py:
```diff
@@ -0,0 +1,93 @@
+		)
+		self.db_set("email_sent_at", frappe.utils.now())
+
+	def after_delete(self):
```
We should "cancel" invitations, not delete them.

ankush commented on frappe/core/doctype/user_invitation/user_invitation.py:
```diff
@@ -0,0 +1,93 @@
+			)
+
+	def on_update(self):
+		if self.has_value_changed("status") and self.status == "Expired":
```
This kind of triggers can be a separate action on the document. E.g. `expire()` or `cancel_invitation()` which updates the doc and does appropriate action.

E.g.

```python
def expire(self):
    self.status = "Expired"
    self.save()
    frappe.sendmail(...)
```

ankush commented on frappe/core/doctype/user_invitation/user_invitation.py:
```diff
@@ -0,0 +1,93 @@
+		if self.has_value_changed("status") and self.status == "Expired":
+			email_title = self.get_email_title()
+			frappe.sendmail(
+				recipients=self.email,
```
We should be notifying the person who sent the invite, not the person who didn't accept.

ankush commented on frappe/core/doctype/user_invitation/user_invitation.py:
```diff
@@ -0,0 +1,93 @@
+				subject=f"Invitation to join {email_title} expired",
+				template="user_invitation_expired",
+				args={"title": email_title, "site_name": self.get_site_name()},
+				now=True,
```
This can be asynchronous, no need for `now=True`.

ankush commented on frappe/core/doctype/user_invitation/user_invitation.py:
```diff
@@ -0,0 +1,93 @@
+			)
+
+	def get_email_title(self):
+		return f"Frappe {(self.app_name if self.app_name != "frappe" else 'framework').capitalize()}"
```
Get this from the hook `app_title`, don't hardcode the naming scheme of apps.

ankush commented on frappe/core/doctype/user_invitation/user_invitation.py:
```diff
@@ -0,0 +1,93 @@
+		email_title = self.get_email_title()
+		frappe.sendmail(
+			recipients=self.email,
+			subject=f"You've been invited to join {email_title}",
```
https://docs.frappe.io/framework/user/en/translations 

Wrap this and other messages to allow translations.

ankush commented on frappe/core/doctype/user_invitation/user_invitation.py:
```diff
@@ -0,0 +1,93 @@
+	for invitation in invitations_to_expire:
+		invitation = frappe.get_doc("User Invitation", invitation.name)
+		invitation.status = "Expired"
+		invitation.save()
```
Add `frappe.db.commit()` in the loop here... because if there are too many "pending" invitations then this will never complete.


https://docs.frappe.io/framework/user/en/api/database#database-transaction-model

ankush commented on frappe/hooks.py:
```diff
@@ -238,6 +238,7 @@
 		"frappe.desk.doctype.event.event.send_event_digest",
 		"frappe.email.doctype.notification.notification.trigger_daily_alerts",
 		"frappe.desk.form.document_follow.send_daily_updates",
+		"frappe.core.doctype.user_invitation.user_invitation.mark_expired_invitations",
```
Move this to `daily_maintenance`

ankush commented on frappe/core/doctype/user_invitation/user_invitation.json:
```diff
@@ -0,0 +1,132 @@
+  "email",
+  "role",
+  "invited_by",
+  "status",
```
Add document states with relevant colors for this.

ankush commented on frappe/core/doctype/user_invitation/user_invitation.json:
```diff
@@ -0,0 +1,132 @@
+ "permissions": [
+  {
+   "create": 1,
+   "delete": 1,
```
Not required. They should be able to cancel instead.

ankush commented on frappe/core/doctype/user_invitation/user_invitation.json:
```diff
@@ -0,0 +1,132 @@
+   "print": 1,
+   "read": 1,
+   "report": 1,
+   "role": "User Invitation Manager",
```
This is largely same as `System Manager`

ankush commented on frappe/core/api/user_invitation.py:
```diff
@@ -0,0 +1,149 @@
+from frappe.model.document import Document
+
+
+def _validate_app_name(app_name: str) -> None:
```
https://github.com/frappe/erpnext/wiki/Coding-Standards#function-sequence 

Follow this rough ordering guideline for readability.

ankush commented on frappe/core/api/user_invitation.py:
```diff
@@ -0,0 +1,149 @@
+	allowed_roles: list[str] = []
+	if app_name != "frappe":
+		allowed_roles = _get_allowed_roles(app_name)
+		_validate_role(role, allowed_roles)
```
These two can be likely merged.

ankush commented on frappe/core/api/user_invitation.py:
```diff
@@ -0,0 +1,149 @@
+
+
+def _validate_emails(emails: str) -> list[str]:
+	frappe.utils.validate_email_address(emails, throw=True)
```
Explicity import `frappe.utils`. This is relying on side-effect of other imports right now.

ankush commented on frappe/core/api/user_invitation.py:
```diff
@@ -0,0 +1,149 @@
+		allowed_roles = _get_allowed_roles(app_name)
+		_validate_role(role, allowed_roles)
+	email_list = _validate_emails(emails)
+	existing_user_emails = _get_existing_user_emails(email_list)
```
For 1/2 line utils, if they're not used more than once, prefer inlining them. It's easier to read then jumping across functions.

ankush commented on frappe/core/api/user_invitation.py:
```diff
@@ -0,0 +1,149 @@
+	existing_invited_emails = _get_existing_invited_emails(
+		email_list, allowed_roles if app_name != "frappe" else None
+	)
+	to_invite = list(set(email_list) - set(existing_user_emails) - set(existing_invited_emails))
```
There should be a simple way to give access to existing users. E.g. Someone might install helpdesk on a site which has ERPNext and all the users already configured.

ankush commented on frappe/core/api/user_invitation.py:
```diff
@@ -0,0 +1,149 @@
+	return key
+
+
+@frappe.whitelist(allow_guest=True)
```
Specify `methods` for all security critical endpoints

ankush commented on frappe/core/api/user_invitation.py:
```diff
@@ -0,0 +1,149 @@
+
+
+def _get_invitation(key: str) -> Document:
+	result = frappe.db.get_all("User Invitation", filters={"key": key}, pluck="name")
```
Prefer `db.get_value` here, directly gives single value or None.

ankush commented on frappe/core/api/user_invitation.py:
```diff
@@ -0,0 +1,149 @@
+	invitation = _get_invitation(key)
+	if invitation.status == "Expired":
+		frappe.throw(_("Invitation is expired"))
+	if invitation.status == "Pending":
```
All the meat of this should be moved on the doctype controller and use doctype methods instead of `_` utils in this file.

ankush commented on frappe/core/api/user_invitation.py:
```diff
@@ -0,0 +1,149 @@
+		user = _upsert_user(invitation.email, invitation.role)
+		_accept_invitation(invitation, user.email)
+		if invitation.app_name != "frappe":
+			_run_after_invite_accept_fns(invitation)
```
Avoid acronyms, short names. Call it `hooks` or `triggers`.

ankush commented on frappe/core/api/user_invitation.py:
```diff
@@ -0,0 +1,149 @@
+	for dot_path in after_accept:
+		after_accept_fns.append(frappe.get_attr(dot_path))
+	for after_accept_fn in after_accept_fns:
+		after_accept_fn(invitation)
```
Use `frappe.call` and `kwargs` to make the hook signature flexible.

ankush commented on frappe/core/api/user_invitation.py:
```diff
@@ -0,0 +1,149 @@
+	)
+
+
+def _set_reset_password_key(user: Document) -> str:
```
See if you can just use whatever `user.py` has instead of duplicating.

ankush commented on frappe/core/api/user_invitation.py:
```diff
@@ -0,0 +1,149 @@
+	# GET requests do not cause an implicit commit
+	frappe.db.commit()  # nosemgrep
+	if not frappe.local.flags.in_test and not should_update_password:
+		frappe.local.login_manager.login_as(invitation.email)
```
Remove the flag and in your test simulate a request using set_request 

https://github.com/frappe/frappe/blob/db0bbea198471179ae26c58819adb461b1b58a9c/frappe/commands/site.py#L1208-L1212

EltonLobo07 replied:
I have used a different workaround to avoid relying on local flags.

ankush reviewed (dismissed):
Nice first major PR :tada:

ankush commented on frappe/core/doctype/user_invitation/test_user_invitation.py:
```diff
@@ -0,0 +1,184 @@
+	@classmethod
+	def setUpClass(cls):
+		super().setUpClass()
+		frappe.set_user("Administrator")
```
This should be teardown

ankush commented:
@EltonLobo07 @shariquerik @RitvikSardana 

If we are adding this to the framework, we should also make it usable from desk. This means:

1. Fix up the form so it's usable by end-users. i.e. Hide internal things, make the form look almost the same as the intended UI in other apps.
2. Enable quick entry mode.

EltonLobo07 committed:
refactor(user-invitation): improve security & readability

Improvements:
- move invite expiration check to `daily_maintenance`
- explicitly import all of the used packages
- specify methods for all security-critical endpoints
- improve error messages and give them suitable titles
- remove unnecessary utility functions
- make invitation key management secure
- translate all of the subjects of the sent emails
- use the `app_title` hook to create email titles
- commit the work done after each iteration of the background invitation
  expiry checker
- restructure code to improve readability
- use `user.reset_password` to generate the target link
- use clear long names to name identifiers
- add document states with relevant colors (User Invitation doctype)
- differ `sendmail` emails whenever possible
- send an email to the invitation creator instead of the invitee after
  the invite has expired
- remove `User Invitation Manager` role

EltonLobo07 committed:
fix(user-invitation): use valid emails to test doctype & related code

EltonLobo07 committed:
feat(user-invitation): support adding multiple roles

EltonLobo07 committed:
refactor(user-invitation): mark relevant fields `set only once`

EltonLobo07 committed:
feat(user-invitation): add `Cancelled` status

EltonLobo07 committed:
test(user-invitation): correct broken tests

EltonLobo07 committed:
test(user-invitation): form valid f-strings & run code formatter

EltonLobo07 committed:
feat(user-invitation): make doctype usable from desk

EltonLobo07 committed:
fix(user-invitation): remove delete permission from invitation doctype

EltonLobo07 committed:
feat(user-invitation): pass user inserted info to `after_accept` hook

EltonLobo07 committed:
refactor(user-invitation): improve custom action methods & errors

Improvements:
- trigger actions only when the invitation is in the `Pending` state
- use lowercase letters to start error messages
- handle cases where `user_invitation_hook` is not defined

EltonLobo07 committed:
refactor(user-invitation): remove site name from email templates

EltonLobo07 committed:
docs(user-invitation): add internal documentation

EltonLobo07 committed:
feat(user-invitation): add 'get pending' & cancel invites apis

EltonLobo07 committed:
fix(user-invitation): make invitation app specific

EltonLobo07 committed:
refactor(user-invitation): avoid mixing function programming

EltonLobo07 committed:
fix(user-invitation): make apis usable for app specific valid users

EltonLobo07 committed:
fix(user-invitation): allow app specific invites

EltonLobo07 committed:
feat(user-invitation): make list view & permission checks app specific

EltonLobo07 committed:
refactor(user-invitation): convert class methods to static when possible

EltonLobo07 committed:
feat(user-invitation): add `app_only_for` method to the doc

EltonLobo07 committed:
fix(user-invitation): f-string syntax error in `get_permission_query_conditions`

EltonLobo07 committed:
docs(user-invitation): add examples & improve the internal doc

ankush committed:
refactor: rename method name

static_ is unnecessary
only_for doesn't make sense in this context when arguments are not roles

ankush committed:
fix: Support POST request too

We dont follow REST semantics 100%, anything that modifies something
should ideally be doable with POST too.

ankush committed:
chore: cap

ankush commented on frappe/core/doctype/user_invitation/user_invitation.py:
```diff
@@ -0,0 +1,241 @@
+		self._run_after_accept_hooks(user, user_inserted)
+
+	@frappe.whitelist()
+	def cancel_invite(self, ignore_permissions: bool = False):
```
This shouldn't be user argument.

ankush committed:
fix: Avoid ignore_permissions as user arg
